File: union-find/java/src/QuickUnion.java
/**
 * Created by petergoldsborough on 08/12/15.
 */

public class QuickUnion extends UnionFind
{
    QuickUnion(int N)
    {
        super(N);
    }
    
    public void union(int p, int q)
    {
        int p_root = _root(p);

        int q_root = _root(q);

        if (p_root != q_root)
        {
            _id[_root(q)] = _root(p);
        }
    }
    
    public boolean connected(int p, int q)
    {
        return _root(p) == _root(q);
    }
    
    protected int _root(int i)
    {
        while(_id[i] != i) i = _id[i];
        
        return i;
    }
}

File: union-find/python/QuickUnion.py
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-

import UnionFind

class QuickUnion(UnionFind.UnionFind):

	def __init__(self, N):
		super().__init__(N)
		self.roots = [x for x in range(N)]

	def union(self, p, q):
		p_root = self.root(p)
		q_root = self.root(q)
		self.roots[q_root] = p_root

	def connected(self, p, q):
		return self.root(p) == self.root(q)

	def root(self, i):
		while i != self.roots[i]:
			i = self.roots[i]
		return i

	def __repr__(self):
		return str(self.roots)

def main():
	UnionFind.main(QuickUnion)

if __name__ == "__main__":
	main()
